src/arena.c

```diff
@@ -462,6 +484,112 @@ arena_decay(tsdn_t *tsdn, arena_t *arena, bool is_background_thread, bool all) {
 	arena_decay_muzzy(tsdn, arena, is_background_thread, all);
 }
 
+static inline uint64_t
+arena_ns_until_purge(tsdn_t *tsdn, decay_t *decay, size_t npages) {
+	if (malloc_mutex_trylock(tsdn, &decay->mtx)) {
+		/* Use minimal interval if decay is contended. */
+		return DEFERRED_ASAP;
+	}
+	uint64_t result = decay_ns_until_purge(decay, npages,
+	    ARENA_DEFERRED_PURGE_NPAGES_THRESHOLD);
+
+	malloc_mutex_unlock(tsdn, &decay->mtx);
+	return result;
+}
+
+/*
+ * Get time until next deferred work ought to happen. If there are multiple
+ * things that have been deferred, this function calculates the time until
+ * the soonest of those things.
+ */
+uint64_t
+arena_time_until_deferred(tsdn_t *tsdn, arena_t *arena) {
+	uint64_t soonest;
+	soonest = arena_ns_until_purge(tsdn,
```

Comment:
I think this logic can be moved into the pa_shard_time_until_differed_word call.

Really, given how often the background thread reaches into PA internals anyways, I'm not sure this function needs to exist at all.

Reply:
Makes sense, will do.

Reply:
Although, while doing this, I found that PA currently doesn't do anything decay-wise. The current implementation puts the responsibility of handling deferred decay on arena, hence having anything related to decay in PA seems a bit off without other changes in deferring decay.

For instance, arena has functions like `arena_dirty_decay_ms_default_get`, `arena_decay`, etc. In my diff, arena is supposed to know `ARENA_DEFERRED_PURGE_NPAGES_THRESHOLD`, and PA is oblivious to this.

Also, can you please elaborate on the frequency of background thread reaching into PA internals? I can see only calls to `pa_shard_set_deferral_allowed`, doesn't seem that often.

How I see it working within previously set boundaries: decay is deferred on arena level, but there can be possibly other stuff deferred by PAC/HPA which we want to add support for. Consequently, backgound_thread would ask each arena for time until deferred work, arena would check itself via `arena_ns_until_purge` and its pa_shard via `pa_shard_time_until_deferred_work`

Reply:
The broad structure we're heading towards is that an arena is a set of bins, some state tracking for large allocations, and a handle to a page allocator shard (which is in turn tied to a pa_central_t). So that the arena's job becomes translating malloc requests to page-level ones. Because historically these were all sort of tied together (the arena was in charge of everything), we've got a lot of places where responsibility for some aspect of the page allocator doesn't really live in the PA module (like with decay, as you note). So e.g. when the background threads / arena access decay_t objects directly (rather than some PA-level wrapper), that's what I mean by "reaching into PA internals" (and similarly, when touching the PA shard ecaches directly).

So when we're adding new functionality that deals with things logically scoped to page management, we should break ties by putting that logic into pa.c or one of the PAI implementations instead of arena.c.

What's called in this diff arena_time_until_deferred should really be pac_time_until_deferred_work (with a little bit of reworking probably; stuff like BACKGROUND_THREAD_DEFERRED_MIN doesn't seem fundamental to the PAC decay logic, it's something that belongs in the caller rather than the callee). The PA shard should have a function that calls pai_time_until_deferred_work on each of its PAI implementations and returns the min (possibly with some adjustment).

Does that make sense? Sorry, I know it's not necessarily super easy to glean that direction from the current state of the codebase.

Reply:
Yes, this direction is clear.

Should I try to move other decay-related stuff from arena into PA?

Reply:
I think that would be a valuable code-cleanup thing, but I wouldn't spend a *lot* of time on it. So if it seems easy to do I'd so go for it, but I wouldn't delay the functional improvements if you start getting bogged down for some unforeseen reason.